File: Problems & Solutions/N/Next Greater Element/16 January 2023/Next Greater Element.cpp
Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    vector<long long> nextLargerElement(vector<long long> &arr, int n){
      vector<long long>ans;
      stack<long long>st;
      for(int i=n-1;i>=0;i--){
      while(!st.empty() && st.top()<=arr[i])st.pop();
      if(st.empty()){
          ans.push_back(-1);
          st.push(arr[i]);
      }
      else{
          ans.push_back(st.top());
          st.push(arr[i]);
      }
      }
      reverse(ans.begin(),ans.end());
      return ans;
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {

        int n;
        cin>>n;
        vector<long long> arr(n);
        for(int i=0;i<n;i++)
            cin>>arr[i];

        Solution ob;

        vector <long long> res = ob.nextLargerElement(arr, n);
        for (long long i : res) cout << i << " ";
        cout<<endl;
    }
	return 0;
}

// } Driver Code Ends

File: Problems & Solutions/N/Next Greater Element/16 January 2023/Next Greater Element.java
//{ Driver Code Starts
/*package whatever //do not write package name here */

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
    
	public static void main (String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine().trim());
		while(t-->0){
		    int n = Integer.parseInt(br.readLine().trim());
		    String inputLine[] = br.readLine().trim().split(" ");
		    long[] arr = new long[n];
		    for(int i=0; i<n; i++)arr[i]=Long.parseLong(inputLine[i]);
		    Solution ob = new Solution();
		    long[] res = ob.nextLargerElement(arr, n);
		    for (int i = 0; i < n; i++) 
		        System.out.print(res[i] + " ");
		    System.out.println();
		}
	}
}




// } Driver Code Ends


// User Function Template for JAVA

class Solution{
    public static long[] nextLargerElement(long[] arr, int n) { 
        
        Stack<Long> st = new Stack<>();
        
        long[] ans = new long[n];
        
        for(int i=n-1;i>=0;i--){
            
            if(st.isEmpty())    ans[i] = -1;
            
            while(!st.isEmpty() && st.peek() <= arr[i]) st.pop();
            
            if(i<n && !st.isEmpty()) ans[i] = st.peek();
            
            else ans[i] = -1;
            
            st.push(arr[i]);
        }
        return ans;
    } 
}
